<a href="https://colab.research.google.com/github/luisfilipe10/DATA_SCIENCE/blob/main/Primeiros%20Passos/Hora_da_pratica.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Sobre o notebook

A "hora da prática" é uma atividade do curso de Python para Data Science da Alura que contém diversos desafios envolvendo os aprendizados de cada aula.

O notebook "hora da pratica" é um notebook construído com espaço para as células dos desafios propostos no curso, com ele você tem espaço para construir e executar suas soluções

## Aula 1 - Introdução ao Python

### Questão 1

Imprima a frase `Escola de Dados da Alura!`

In [ ]:
print('Escola de Dados da Alura')

Escola de Dados da Alura


### Questão 2
Imprima na tela seu nome e seu sobrenome seguindo a estrutura abaixo:
```
Nome: [seu nome]
Sobrenome: [seu sobrenome]
```


In [ ]:
print("Nome: Luis")

Nome: Luis


In [ ]:
print("Sobrenome: Martins")

Sobrenome: Martins


### Questão 3
Imprima o seu primeiro nome letra a letra. Por exemplo, nome é Mirla, então eu obtenho a seguinte saída:
```
M
I
R
L
A
```


### Questão 4
Imprima o dia do seu nascimento em formato `dia mês ano`. Os valores de dia e ano não podem estar entre aspas. O formato deve estar como no exemplo, supondo uma data de aniversário dia 28 de fevereiro de 2003:
```
28 fevereiro 2003
```

In [ ]:
print(28,'Fevereiro',2003)

28 Fevereiro 2003


### Questão 5
Imprima em um único print o atual ano que você está fazendo esse curso, o valor do ano deve ser um dado numérico. A saída do print deve ser a seguinte:
```
Ano atual: [ano]
```


In [ ]:
print('Ano atual:', 2025)

Ano atual: 2025


## Aula 2 - Manipulando dados no Python

### Coleta e amostragem de dados

#### Questão 1
Crie um programa que solicite à pessoa usuária digitar seu nome, e imprima “Olá, [nome]!”.

In [ ]:
nome = input('Digite seu nome: ')
print('Olá,', nome)

Digite seu nome: luis
Olá, luis


#### Questão 2
Crie um programa que solicite à pessoa usuária digitar seu nome e idade, e imprima “Olá, [nome], você tem [idade] anos.”.

In [ ]:
info= input('Digite seu nome e idade: ')
print('Olá,', info)

Digite seu nome e idade: luis 2
Olá, luis 2


#### Questão 3
Crie um programa que solicite à pessoa usuária digitar seu nome, idade e altura em metros, e imprima “Olá, [nome], você tem [idade] anos e mede [altura] metros!”.

In [ ]:
nome = input('Digite seu nome:')
idade = input('Digite sua idade:')
altura = input('Digite sua altura:')
print('Olá,', nome, 'você tem', idade, 'anos e mede', altura, 'metros!')


Digite seu nome:luis
Digite sua idade:20
Digite sua altura:1.82
Olá, luis você tem 20 anos e mede 1.82 metros!


### Calculadora com operadores

#### Questão 1

Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a soma dos dois valores.

In [ ]:
num1 = int(input('Insira um numero aleatorio: '))
num2 = int(input('Insiraa um numero aleatorio: '))
print(num1+num2)

Insira um numero aleatorio: 2
Insiraa um numero aleatorio: 3
5


#### Questão 2
Crie um programa que solicite três valores numéricos à pessoa usuária e imprima a soma dos três valores.

In [ ]:
num1= int(input('Insira um numero aleatorio: '))
num2= int(input('Insira um numero aleatorio: '))
num3= int(input('Insira um numero aleatorio: '))
print(num1+num2+num3)


Insira um numero aleatorio: 3
Insira um numero aleatorio: 5
Insira um numero aleatorio: 2
10


In [ ]:
soma = 0

for i in range(3):
  numero = int(input('Insira seu numero: '))
  soma += numero

print(soma)

Insira seu numero: 3
Insira seu numero: 2
Insira seu numero: 7
12


#### Questão 3
Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a subtração do primeiro pelo o segundo valor.

In [ ]:
numero1 = int(input('Insira um numero aleatorio: '))
numero2 = int(input('Insira um numero aleatorio: '))
print(numero1-numero2)
#

Insira um numero aleatorio: 33
Insira um numero aleatorio: 2
31


#### Questão 4

Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a multiplicação dos dois valores.


In [ ]:
num1 = float(input('Insira um numero aleatorio: '))
num2 = float(input('Insira um numero aleatorio: '))
print(num1*num2)
#

Insira um numero aleatorio: 3.1
Insira um numero aleatorio: 66
204.6


#### Questão 5

Crie um programa que solicite dois valores numéricos, um numerador e um denominador, e realize a divisão entre os dois valores. Deixe claro que o valor do denominador não pode ser `0`.


In [ ]:
num = float(input('Insira seu denominador: '))
den = float(input('Insira seu numerador: '))
print(num/den)

Insira seu denominador: 10
Insira seu numerador: 2
5.0


#### Questão 6

Crie um programa que solicite dois valores numéricos, um operador e uma potência, e realize a exponenciação entre esses dois valores.


In [ ]:
num = float(input('Insira seu numerador: '))
pot = float(input('Insira sua potencia: '))

print(num**pot)

Insira seu numerador: 2.5
Insira sua potencia: 5.2
117.29730800599916


#### Questão 7
Crie um programa que solicite dois valores numéricos, um numerador e um denominador e realize a divisão inteira entre os dois valores. Deixe claro que o valor do denominador não pode ser `0`.

In [ ]:
num = float(input('Insira seu numerador: '))
den = float(input('Insira seu denominador: '))
print(num//den)

Insira seu numerador: 2
Insira seu denominador: 15
0.0


#### Questão 8

Crie um programa que solicite dois valores numéricos, um numerador e um denominador, e retorne o resto da divisão entre os dois valores. Deixe claro que o valor do denominador não pode ser `0`.

In [ ]:
den= float(input('Insira seu denominador: '))
num= float(input('Insira seu numerador: '))
print(num%den)

Insira seu denominador: 2
Insira seu numerador: 1
1.0


#### Questão 9

Crie um código que solicita 3 notas de um estudante e imprima a média das notas.

In [ ]:
soma = 0
for i in range(3):
  nota = float(input('Insira sua nota: '))
  soma += nota
print((soma/3))
print(soma)

Insira sua nota: 3
Insira sua nota: 5
Insira sua nota: 6
4.666666666666667
14.0


#### Questão 10

Crie um código que calcule e imprima a média ponderada dos números 5, 12, 20 e 15 com pesos respectivamente iguais a 1, 2, 3 e 4.

In [ ]:
numeros = [5, 12, 20, 15]
pesos = [1, 2, 3, 4]

soma_ponderada = 0
soma_pesos = 0

for i in range(len(numeros)):
    soma_ponderada += numeros[i] * pesos[i]
    soma_pesos += pesos[i]

media_ponderada = soma_ponderada / soma_pesos

print(f"A média ponderada é: {media_ponderada}")

A média ponderada é: 14.9


### Editando textos

#### Questão 1

Crie uma variável chamada “`frase`” e atribua a ela uma string de sua escolha. Em seguida, imprima a frase na tela.

In [ ]:
frase = input('Insira algo: ')
print(frase)

Insira algo: tudo nosso
tudo nosso


#### Questão 2
Crie um código que solicite uma frase e depois imprima a frase na tela.

In [ ]:
frasee = input('Inmsira uma frase: ')
print(frasee)

Inmsira uma frase: asdasd
asdasd


#### Questão 3

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase digitada mas com todas as letras maiúsculas.

In [ ]:
frase = input('Insira uma frase aleatoria: ')
print(frase.upper())

Insira uma frase aleatoria: adasdasd
ADASDASD


#### Questão 4

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase digitada mas com todas as letras minúsculas.

In [ ]:
frase = input('Insira uma frase aleatoria: ')
print(frase.lower())

Insira uma frase aleatoria: ADDFASD
addfasd


#### Questão 5

Crie uma variável chamada “`frase`” e atribua a ela uma string de sua escolha. Em seguida, imprima a frase sem espaços em branco no início e no fim.

In [ ]:
frase = input('Insira uma frase aleatoria: ')
print(frase.strip())

Insira uma frase aleatoria: e rer    rerrr
e rer    rerrr


#### Questão 6

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase sem espaços em branco no início e no fim.

In [ ]:
frase = input("Digite uma frase: ")

frase_sem_espacos = frase.strip()

print(frase_sem_espacos)

Digite uma frase: asda daasd a3ewef ff    fff
asda daasd a3ewef ff    fff


#### Questão 7


Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase sem espaços em branco no início e no fim e em letras minúsculas.

In [ ]:
frase = input("Digite uma frase: ")

frase_sem_espacos = frase.strip()

print(frase_sem_espacos)


#### Questão 8

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase com todas as vogais “`e`” trocadas pela letra “`f`”.

In [ ]:
frase = input("Digite uma frase: ")

frase_trocada = frase.replace('e', 'f')

print(frase_trocada)

Digite uma frase: ola munde tudo nosse
ola mundf tudo nossf


In [ ]:
frase = input('insira um texto: ')
print(frase.replace('e','f'))

insira um texto: nosse
nossf


#### Questão 9


Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase com todas as vogais “`a`” trocadas pela caractere  “`@`”.

In [ ]:
frase = input('Insira sua frase: ')
print(frase.replace('a', '@'))

Insira sua frase: alasfa
@l@sf@


#### Questão 10


Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase com todas as consoantes “`s`” trocadas pelo caractere  “`$`”.

In [ ]:
frase = input('Insira sua frase: ')
print(frase.replace('s', '$'))

Insira sua frase: asdass
a$da$$


## Aula 3 - Estruturas condicionais

### Aquecendo na programação

#### Questão 1
Escreva um programa que peça à pessoa usuária para fornecer dois números e exibir o número maior.

In [5]:
num1 = input('Insira um numero: ')
num2 = input('Insira um numero: ')
print('o maior eh: ', max(num1, num2))

Insira um numero: 2
Insira um numero: 3
o maior eh:  3


In [3]:
num1 = input('Insira um numero: ')
num2 = input('Insira um numero: ')

if num1 > num2:
  print(f'o numero maior eh {num1}')
else:
  print(f'o numero maior eh {num2}')

Insira um numero: 33
Insira um numero: 
o numero maior eh 33


#### Questão 2
Escreva um programa que solicite o percentual de crescimento de produção de uma empresa e informe se houve um crescimento (porcentagem positiva) ou decrescimento (porcentagem negativa).

In [7]:
percentil = float(input('Insira o percentual de producao nesse mes: '))
if percentil > 0 :
  print('Houve um crescimento na producao!! ')
elif percentil < 0 :
  print('Houve um decrescimento na producao!!')
else:
  print('Houve um crescimento nulo!!')

Insira o percentual de producao nesse mes: 1
Houve um crescimento na producao!! 


#### Questão 3
Escreva um programa que determine se uma letra fornecida pela pessoa usuária é uma vogal ou consoante.

In [ ]:
# prompt: Escreva um programa que determine se uma letra fornecida pela pessoa usuária é uma vogal ou consoante, maais simples

letra = input("Digite uma letra: ").lower()

if letra in "aeiou":
  print("A letra é uma vogal.")
elif letra.isalpha():
  print("A letra é uma consoante.")
else:
  print("Caractere inválido. Digite uma letra.")


In [12]:
letra = input('Insira umaa letra: ').strip().lower()

if letra in 'aeiou':
  print('eh uma vogal')
else:
  print('eh uma consoante')

Insira umaa letra: a
eh uma vogal


#### Questão 4
Escreva um programa que leia valores médios de preços de um modelo de carro por 3 anos consecutivos e exiba o valor mais alto e mais baixo entre esses três anos.

In [ ]:
# prompt: Escreva um programa que leia valores médios de preços de um modelo de carro por 3 anos consecutivos e exiba o valor mais alto e mais baixo entre esses três anos. usando estruturaa de repeticao

maior_preco = float('-inf')  # Inicializa com um valor negativo infinito
menor_preco = float('inf')   # Inicializa com um valor positivo infinito

for ano in range(3):
  preco = float(input(f"Digite o preço médio do carro no ano {ano + 2023}: "))
  if preco > maior_preco:
    maior_preco = preco
  if preco < menor_preco:
    menor_preco = preco

print(f"O maior preço médio foi: R${maior_preco:.2f}")
print(f"O menor preço médio foi: R${menor_preco:.2f}")


In [31]:


for valor in range(3) :
  valor = input('Insira  o valor ddp preco do carro: ')
print('o valor mais alto eh : ',(int(max(valor))))


Insira  o valor ddp preco do carro: 10000
Insira  o valor ddp preco do carro: 30000
Insira  o valor ddp preco do carro: 3000
o valor mais alto eh :  3


In [ ]:
preco = 0

for carro in range(2017,2020)



#### Questão 5
Escreva um programa que pergunte sobre o preço de três produtos e indique qual é o produto mais barato para comprar.

In [29]:


for produto in range(3):
  preco = input('Insira o preco do produto: ')
print('o produto mais barato eh: ',(min(preco)))



Insira o preco do produto: 5
Insira o preco do produto: 66
Insira o preco do produto: 3
o produto mais barato eh:  3


#### Questão 6
Escreva um programa que leia três números e os exiba em ordem decrescente.

In [ ]:
# prompt: Escreva um programa que leia três números e os exiba em ordem decrescente, dde maneira maais simples

numeros = []
for i in range(3):
  numero = int(input("Digite um número: "))
  numeros.append(numero)

numeros.sort(reverse=True)
print("Números em ordem decrescente:", numeros)


#### Questão 7
Escreva um programa que pergunte em qual turno a pessoa usuária estuda ("manhã", "tarde" ou "noite") e exiba a mensagem "Bom Dia!", "Boa Tarde!", "Boa Noite!", ou "Valor Inválido!", conforme o caso.

In [36]:
horario = input('Insira seu horario de aula: ')

if 'manha' in horario:
  print('Bom dia')
elif 'tarde' in horario:
  print('Boa tarde')
elif 'noite' in horario:
  print('Boa noite')
else:
  print('Valor invalido')


Insira seu horario de aula: noite
Boa noite


#### Questão 8
Escreva um programa que peça um número inteiro à pessoa usuária e determine se ele é par ou ímpar. Dica: Você pode utilizar o operador módulo `%`.

In [ ]:
# prompt: Escreva um programa que peça um número inteiro à pessoa usuária e determine se ele é par ou ímpar. Dica: Você pode utilizar o operador módulo %.

numero = int(input("Digite um número inteiro: "))

if numero % 2 == 0:
  print(f"O número {numero} é par.")
else:
  print(f"O número {numero} é ímpar.")


In [39]:
numero = int(input('Insira um numero: '))
if numero % 2 == 0:
  print('eh par')
else:
  print('eh impar')

Insira um numero: 4
eh par


#### Questão 9
Escreva um programa que peça um número à pessoa usuária e informe se ele é inteiro ou decimal.

In [ ]:
# prompt: Escreva um programa que peça um número à pessoa usuária e informe se ele é inteiro ou decimal.

numero = input("Digite um número: ")

try:
  numero_float = float(numero)
  if numero_float.is_integer():
    print("O número é inteiro.")
  else:
    print("O número é decimal.")
except ValueError:
  print("Entrada inválida. Por favor, digite um número válido.")


### Momento dos projetos

#### Questão 10
Um programa deve ser escrito para ler dois números e, em seguida, perguntar à pessoa usuária qual operação ele deseja realizar. O resultado da operação deve incluir informações sobre o número - se é par ou ímpar, positivo ou negativo e inteiro ou decimal.

In [42]:
num1 = float(input('Insira um numero: '))
num2 = float(input('Insira outro numero: '))

operacao = input('qual operacao deseja fazer?: ')

if operacao == 'soma':
  resultado = num1 + num2
elif operacao == 'subtracao':
  resultado = num1 - num2
elif operacao == 'multiplicacao':
  resultado = num1 * num2
elif operacao == 'divisao':
  resultado = num1 / num2
print(resultado)

if resultado % 2 ==0:
  print('eh par')
else:
  print('eh impar')

if resultado > 0:
  print('eh positivo')
else:
  print('eh negativo')

Insira um numero: 3
Insira outro numero: 2
qual operacao deseja fazer?: soma
5.0
eh impar
eh positivo


#### Questão 11
Escreva um programa que peça à pessoa usuária três números que representam os lados de um triângulo. O programa deve informar se os valores podem ser utilizados para formar um triângulo e, caso afirmativo, se ele é equilátero, isósceles ou escaleno. Tenha em mente algumas dicas:
- Três lados formam um triângulo quando a soma de quaisquer dois lados for maior que o terceiro;
- Triângulo Equilátero: três lados iguais;
- Triângulo Isósceles: quaisquer dois lados iguais;
- Triângulo Escaleno: três lados diferentes;


In [43]:
# prompt: Escreva um programa que peça à pessoa usuária três números que representam os lados de um triângulo. O programa deve informar se os valores podem ser utilizados para formar um triângulo e, caso afirmativo, se ele é equilátero, isósceles ou escaleno. Tenha em mente algumas dicas:
# Três lados formam um triângulo quando a soma de quaisquer dois lados for maior que o terceiro;
# Triângulo Equilátero: três lados iguais;
# Triângulo Isósceles: quaisquer dois lados iguais;
# Triângulo Escaleno: três lados diferentes;

lado1 = float(input("Digite o comprimento do primeiro lado: "))
lado2 = float(input("Digite o comprimento do segundo lado: "))
lado3 = float(input("Digite o comprimento do terceiro lado: "))

if lado1 + lado2 > lado3 and lado1 + lado3 > lado2 and lado2 + lado3 > lado1:
  print("Os lados formam um triângulo!")

  if lado1 == lado2 == lado3:
    print("O triângulo é equilátero.")
  elif lado1 == lado2 or lado1 == lado3 or lado2 == lado3:
    print("O triângulo é isósceles.")
  else:
    print("O triângulo é escaleno.")

else:
  print("Os lados não podem formar um triângulo.")


KeyboardInterrupt: Interrupted by user

#### Questão 12
Um estabelecimento está vendendo combustíveis com descontos variados. Para o etanol, se a quantidade comprada for até 15 litros, o desconto será de 2% por litro. Caso contrário, será de 4% por litro. Para o diesel, se a quantidade comprada for até 15 litros, o desconto será de 3% por litro. Caso contrário, será de 5% por litro. O preço do litro de diesel é R\$ 2,00 e o preço do litro de etanol é R\$ 1,70. Escreva um programa que leia a quantidade de litros vendidos e o tipo de combustível (E para etanol e D para diesel) e calcule o valor a ser pago pelo cliente. Tenha em mente algumas dicas:

- O do valor do desconto será a multiplicação entre preço do litro, quantidade de litros e o valor do desconto.
- O valor a ser pago por um cliente será o resultado da multiplicação do preço do litro pela quantidade de litros menos o valor de desconto resultante do cálculo.


In [44]:
# prompt: Um estabelecimento está vendendo combustíveis com descontos variados. Para o etanol, se a quantidade comprada for até 15 litros, o desconto será de 2% por litro. Caso contrário, será de 4% por litro. Para o diesel, se a quantidade comprada for até 15 litros, o desconto será de 3% por litro. Caso contrário, será de 5% por litro. O preço do litro de diesel é R$ 2,00 e o preço do litro de etanol é R$ 1,70. Escreva um programa que leia a quantidade de litros vendidos e o tipo de combustível (E para etanol e D para diesel) e calcule o valor a ser pago pelo cliente. Tenha em mente algumas dicas:
# O do valor do desconto será a multiplicação entre preço do litro, quantidade de litros e o valor do desconto.
# O valor a ser pago por um cliente será o resultado da multiplicação do preço do litro pela quantidade de litros menos o valor de desconto resultante do cálculo.

quantidade_litros = float(input("Digite a quantidade de litros vendidos: "))
tipo_combustivel = input("Digite o tipo de combustível (E para etanol, D para diesel): ").upper()

preco_etanol = 1.70
preco_diesel = 2.00

if tipo_combustivel == 'E':
  if quantidade_litros <= 15:
    desconto = 0.02
  else:
    desconto = 0.04
  valor_desconto = preco_etanol * quantidade_litros * desconto
  valor_a_pagar = (preco_etanol * quantidade_litros) - valor_desconto
  print(f"O valor a ser pago é R$ {valor_a_pagar:.2f}")
elif tipo_combustivel == 'D':
  if quantidade_litros <= 15:
    desconto = 0.03
  else:
    desconto = 0.05
  valor_desconto = preco_diesel * quantidade_litros * desconto
  valor_a_pagar = (preco_diesel * quantidade_litros) - valor_desconto
  print(f"O valor a ser pago é R$ {valor_a_pagar:.2f}")
else:
  print("Tipo de combustível inválido.")


Digite a quantidade de litros vendidos: 100
Digite o tipo de combustível (E para etanol, D para diesel): E
O valor a ser pago é R$ 163.20


#### Questão 13
Em uma empresa de venda de imóveis você precisa criar um código que analise os dados de vendas anuais para ajudar a diretoria na tomada de decisão. O código precisa coletar os dados de quantidade de venda durante os anos de 2022 e 2023 e fazer um cálculo de variação percentual. A partir do valor da variação, deve ser enviada às seguintes sugestões:
- Para variação acima de 20%: bonificação para o time de vendas.
- Para variação entre 2% e 20%: pequena bonificação para time de vendas.
- Para variação entre 2% e -10%: planejamento de políticas de incentivo às vendas.
- Para bonificações abaixo de -10%: corte de gastos.


In [45]:
6# prompt: Em uma empresa de venda de imóveis você precisa criar um código que analise os dados de vendas anuais para ajudar a diretoria na tomada de decisão. O código precisa coletar os dados de quantidade de venda durante os anos de 2022 e 2023 e fazer um cálculo de variação percentual. A partir do valor da variação, deve ser enviada às seguintes sugestões:
# Para variação acima de 20%: bonificação para o time de vendas.
# Para variação entre 2% e 20%: pequena bonificação para time de vendas.
# Para variação entre 2% e -10%: planejamento de políticas de incentivo às vendas.
# Para bonificações abaixo de -10%: corte de gastos.

vendas_2022 = float(input("Digite a quantidade de vendas em 2022: "))
vendas_2023 = float(input("Digite a quantidade de vendas em 2023: "))

variacao_percentual = ((vendas_2023 - vendas_2022) / vendas_2022) * 100

print(f"A variação percentual de vendas entre 2022 e 2023 foi de {variacao_percentual:.2f}%")

if variacao_percentual > 20:
  print("Sugestão: Bonificação para o time de vendas.")
elif 2 <= variacao_percentual <= 20:
  print("Sugestão: Pequena bonificação para o time de vendas.")
elif -10 <= variacao_percentual < 2:
  print("Sugestão: Planejamento de políticas de incentivo às vendas.")
else:
  print("Sugestão: Corte de gastos.")


Digite a quantidade de vendas em 2022: 3
Digite a quantidade de vendas em 2023: 5
A variação percentual de vendas entre 2022 e 2023 foi de 66.67%
Sugestão: Bonificação para o time de vendas.


## Aula 4 - Estruturas de repetição

### Aquecendo na programação

#### Questão 1
Escreva um programa que peça dois números inteiros e imprima todos os números inteiros entre eles.

In [48]:
for i in range(1,2):
  numero1 = float(input('Insira um numero: '))
  numero2 = float(input('Insira outro numero: '))
print(numero1, numero2)

Insira um numero: 2
Insira outro numero: 5
2.0 5.0


#### Questão 2
Escreva um programa para calcular quantos dias levará para a colônia de uma bactéria A ultrapassar ou igualar a colônia de uma bactéria B, com base nas taxas de crescimento de 3% e 1,5% respectivamente. Considere que a colônia A inicia com 4 elementos e a colônia B com 10 elementos.

In [ ]:
# prompt: Escreva um programa para calcular quantos dias levará para a colônia de uma bactéria A ultrapassar ou igualar a colônia de uma bactéria B, com base nas taxas de crescimento de 3% e 1,5% respectivamente. Considere que a colônia A inicia com 4 elementos e a colônia B com 10 elementos.

colonia_a = 4
colonia_b = 10
dias = 0

while colonia_a < colonia_b:
  colonia_a *= 1.03
  colonia_b *= 1.015
  dias += 1

print(f"Levará {dias} dias para a colônia A ultrapassar ou igualar a colônia B.")


In [53]:
colA = 4
colB = 10
dias = 0

while colA <= colB:
  colA *= 1.03
  colB *= 1.01
  dias +=1
print(f"Levará {dias} dias para a colônia A ultrapassar ou igualar a colônia B.")

Levará 47 dias para a colônia A ultrapassar ou igualar a colônia B.


In [54]:
colA = 4
colB = 10
dias = 0

while colA <= colB:
  colA *= 1.03
  colB *= 1.01
  dias +=1
  print(colA, colB)

4.12 10.1
4.2436 10.201
4.370908 10.30301
4.5020352400000005 10.4060401
4.6370962972 10.510100501
4.7762091861160005 10.615201506010001
4.919495461699481 10.721353521070101
5.0670803255504655 10.828567056280802
5.21909273531698 10.93685272684361
5.375665517376489 11.046221254112046
5.536935482897784 11.156683466653167
5.703043547384717 11.268250301319698
5.874134853806259 11.380932804332895
6.050358899420448 11.494742132376224
6.231869666403061 11.609689553699987
6.418825756395154 11.725786449236987
6.611390529087008 11.843044313729356
6.809732244959618 11.96147475686665
7.014024212308407 12.081089504435317
7.22444493867766 12.20190039947967
7.4411782868379905 12.323919403474468
7.664413635443131 12.447158597509214
7.894346044506425 12.571630183484306
8.131176425841618 12.69734648531915
8.375111718616866 12.824319950172342
8.626365070175373 12.952563149674065
8.885156022280635 13.082088781170805
9.151710702949053 13.212909668982514
9.426262024037525 13.345038765672339
9.70904988475865 

#### Questão 3
Para tratar uma quantidade de 15 dados de avaliações de pessoas usuárias de um serviço da empresa, precisamos verificar se as notas são válidas. Então, escreva um programa que vai receber a nota de 0 a 5 de todos os dados e verificar se é um valor válido. Caso seja inserido uma nota acima de 5 ou abaixo de 0, repita até que a pessoa usuária insira um valor válido.

In [66]:
contador = 1

for i in range(0,15):
  print(contador)
  nota = int(input('insira uma nota:' ))

  contador += 1

  if nota > 5 or nota < 0:
    print('nota invalida')
  else:
    print('nota valida')

1
insira uma nota:3
nota valida
2
insira uma nota:2
nota valida
3
insira uma nota:5
nota valida
4
insira uma nota:7
nota invalida
5
insira uma nota:32
nota invalida
6
insira uma nota:4
nota valida
7
insira uma nota:3
nota valida
8
insira uma nota:2
nota valida
9
insira uma nota:9
nota invalida
10
insira uma nota:4
nota valida
11
insira uma nota:3
nota valida
12
insira uma nota:2
nota valida
13
insira uma nota:4
nota valida
14
insira uma nota:2
nota valida
15
insira uma nota:2
nota valida


#### Questão 4
Desenvolva um programa que leia um conjunto indeterminado de temperaturas em Celsius e faça uma análise. Portanto, escreva um programa que leia temperaturas e informe a média delas. A leitura deve ser encerrada ao ser enviado o valor -273°C.

#### Questão 5
Escreva um programa que calcule o fatorial de um número inteiro fornecido pela pessoa usuária. O fatorial de um número inteiro é a multiplicação desse número por todos os seus antecessores até o número 1. Por exemplo, o fatorial de 5 é 5 x 4 x 3 x 2 x 1 = 120.

### Momento dos projetos

#### Questão 6
Escreva um programa que gere a tabuada de um número inteiro de 1 a 10, de acordo com a escolha da pessoa usuária. Como exemplo, para o número 2, a tabuada deve ser mostrada no seguinte formato:
```
Tabuada do 2:
2 x 1 = 2
2 x 2 = 4
[...]
2 x 10 = 20
```

#### Questão 7
Os números primos possuem várias aplicações dentro da Ciência de Dados, por exemplo, na criptografia e segurança. Um número primo é aquele que é divisível apenas por um e por ele mesmo. Faça um programa que peça um número inteiro e determine se ele é ou não um número primo.

#### Questão 8
Vamos entender a distribuição de idades de pensionistas de uma empresa de previdência. Escreva um programa que leia as idades de uma quantidade não informada de clientes e mostre a distribuição em intervalos de [0-25], [26-50], [51-75] e [76-100]. Encerre a entrada de dados com um número negativo.

#### Questão 9
Em uma eleição para gerência em uma empresa com 20 funcionários, existem quatro candidatos. Escreva um programa que calcule o vencedor da eleição. A votação ocorreu da seguinte maneira:
- Cada funcionário votou em um dos quatro candidatos (representados pelos números 1, 2, 3 e 4).
- Também foram contabilizados os votos nulos (representado pelo número 5) e os votos em branco (representado pelo número 6).

Ao final da votação, o programa deve exibir o total de votos para cada candidato, o número de votos nulos e o número de votos em branco. Além disso, deve calcular e exibir a porcentagem de votos nulos em relação ao total de votos e a porcentagem de votos em branco em relação ao total de votos.

## Aula 5 - Estruturas de dados

### Aquecendo na programação

#### Questão 1
Faça um programa que tenha a seguinte lista contendo os valores de gastos de uma empresa de papel `[2172.54, 3701.35, 3518.09, 3456.61, 3249.38, 2840.82, 3891.45, 3075.26, 2317.64, 3219.08]`. Com esses valores, faça um programa que calcule a média de gastos. Dica, use as funções built-in [`sum()`](https://docs.python.org/3/library/functions.html#sum) e [`len()`](https://docs.python.org/3/library/functions.html#len).

#### Questão 2
 Com os mesmos dados da questão anterior, defina quantas compras foram acima de 3000 reais e calcule a porcentagem quanto ao total de compras.

#### Questão 3
Faça um código que colete em uma lista 5 números inteiros quaisquer e imprima a lista. Exemplo: `[1,4,7,2,4]`.

#### Questão 4
Colete novamente 5 inteiros e imprima a lista em ordem inversa à enviada.

#### Questão 5
Faça um programa que, ao inserir um número qualquer, criará uma lista contendo todos os números primos entre 1 e o número digitado.

#### Questão 6
Escreva um programa que peça uma data informando o dia, mês e ano e determine se ela é uma data válida para uma análise.

### Momento dos projetos

#### Questão 7
Para um estudo envolvendo o nível de multiplicação de bactérias em uma colônia, foi coletado o número de bactérias multiplicadas por dia, com base em um processo de duplicação diária e pode ser observado a seguir: `[1.2, 2.1, 3.3, 5.0, 7.8, 11.3, 16.6, 25.1, 37.8, 56.9]`. Tendo esses valores, faça um código que gere uma lista contendo o percentual de crescimento de bactérias por dia, comparando o número de bactérias em cada dia com o número de bactérias do dia anterior. Dica: para calcular o percentual de crescimento usamos a seguinte equação: `100 * (amostra_atual - amostra_passada) / (amostra_passada)`.

#### Questão 8
Para uma seleção de produtos alimentícios, precisamos separar o conjunto de IDs dados por números inteiros, sabendo que os produtos com ID par são doces e os com ID ímpar são amargos. Monte um código que colete 10 IDs. Depois, calcule e mostre a quantidade de produtos doces e amargos.

#### Questão 9
Desenvolva um programa que informa a nota de um aluno de acordo com suas respostas. Ele deve pedir a resposta de um aluno para cada questão e é preciso verificar se a resposta foi igual ao gabarito. Cada questão vale um ponto e existem alternativas:  A, B, C ou D.
```
Gabarito da Prova:
01 - D
02 - A
03 - C
04 - B
05 - A
06 - D
07 - C
08 - C
09 - A
10 - B
```

#### Questão 10
Um instituto de meteorologia deseja fazer um estudo de temperatura média de cada mês do ano. Para isso, você precisa fazer um código que colete e armazene essas temperaturas médias em uma lista. Depois, calcule a média anual das temperaturas e mostre todas as temperaturas acima da média anual, e em que mês elas ocorreram, mostrando os meses por extenso: Janeiro, Fevereiro, etc.

#### Questão 11
Uma empresa de e-commerce está interessada em analisar as vendas dos seus produtos. Os dados das vendas foram armazenados em um dicionário:
```
{'Produto A': 300, 'Produto B': 80, 'Produto C': 60,
 'Produto D': 200, 'Produto E': 250, 'Produto F': 30}
```
Escreva um código que calcule o total de vendas e o produto mais vendido.

#### Questão 12
Uma pesquisa de mercado foi feita para decidir qual design de uma marca infantil mais agrada crianças. A pesquisa foi feita e o votos computados podem ser observados abaixo:
```
'''
Tabela de votos da marca
Design 1 - 1334 votos
Design 2 - 982 votos
Design 3 - 1751 votos
Design 4 - 210 votos
Design 5 - 1811 votos
'''
```
Adapte os dados fornecidos a você para uma estrutura de dicionário e a partir dele, informe o design vencedor e a porcentagem de votos recebidos.


#### Questão 13
Os funcionários de um setor da empresa que você trabalha vão receber um abono correspondente a 10% do seu salário devido ao ótimo desempenho do time. O setor financeiro solicitou sua ajuda para a verificação das consequências financeiras que esse abono irá gerar nos recursos. Assim, foi encaminhada para você uma lista com os salários que receberão o abono: `[1172, 1644, 2617, 5130, 5532, 6341, 6650, 7238, 7685, 7782, 7903]`. O abono de cada funcionário não pode ser inferior a 200. Em código, transforme cada um dos salários em chaves de um dicionário e o abono de cada salário no elemento. Depois, informe o total de gastos com o abono, quantos funcionários receberam o abono mínimo e qual o maior valor de abono fornecido.


#### Questão 14
Uma equipe de cientistas de dados está estudando a diversidade biológica em uma floresta. A equipe fez a coleta de informações sobre o número de espécies de plantas e animais em cada área da floresta e armazenaram essas informações em um dicionário. Nele, a chave descreve a área dos dados e os valores nas listas correspondem às espécies de plantas e animais nas áreas, respectivamente.
```
{'Área Norte': [2819, 7236],
 'Área Leste': [1440, 9492],
 'Área Sul': [5969, 7496],
 'Área Oeste': [14446, 49688],
 'Área Centro': [22558, 45148]}
```
Escreva um código para calcular a média de espécies por área e identificar a área com a maior diversidade biológica. Dica: use as funções built-in [`sum()`](https://docs.python.org/3/library/functions.html#sum) e [`len()`](https://docs.python.org/3/library/functions.html#len).


#### Questão 15
O setor de RH da sua empresa te pediu uma ajuda para analisar as idades dos funcionários de 4 setores da empresa. Para isso, ele te forneceu os seguintes dados:
```
{'Setor A': [22, 26, 30, 30, 35, 38, 40, 56, 57, 65],
 'Setor B': [22, 24, 26, 33, 41, 49, 50, 54, 60, 64],
 'Setor C': [23, 26, 26, 29, 34, 35, 36, 41, 52, 56],
 'Setor D': [19, 20, 25, 27, 34, 39, 42, 44, 50, 65]}
```
Sabendo que cada setor tem 10 funcionários, construa um código que calcule a média de idade de cada setor, a idade média geral entre todos os setores e quantas pessoas estão acima da idade média geral.
